src/odemis/driver/test/tfsbc_test.py

```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
+"""
+Created on 11 May 2020
+
+@author: Philip Winkler
+
+Copyright © 2020 Philip Winkler, Delmic
+
+This file is part of Odemis.
+
+Odemis is free software: you can redistribute it and/or modify it under the terms
+of the GNU General Public License version 2 as published by the Free Software
+Foundation.
+
+Odemis is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+Odemis. If not, see http://www.gnu.org/licenses/.
+"""
+
+from odemis.driver import tfsbc
+from odemis import model
+import unittest
+import logging
+import os
+import time
+import math
+
+TEST_NOHW = (os.environ.get("TEST_NOHW", 0) != 0)  # Default to Hw testing
+
+logging.getLogger().setLevel(logging.DEBUG)
+logging.basicConfig(format="%(asctime)s  %(levelname)-7s %(module)s:%(lineno)d %(message)s")
+
+if TEST_NOHW:
+    PORT = "/dev/fake"
+else:
+    PORT = "/dev/ttyACM*"
+
+# @skip("skip")
+class TestBeamController(unittest.TestCase):
+    """
+    Tests the beam controller driver.
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        cls.bc = tfsbc.BeamController("DC Offset", None, PORT)
+        cls.bc._metadata[model.MD_CALIB_SCALE] = [3e6, 3e6]  # 3e6 amp/m = 3 mamp / nm
+        cls.bc._metadata[model.MD_CALIB_TRANSLATION] = [0, 0]
+        cls.bc._metadata[model.MD_CALIB_ROTATION] = 0  # rad
+
+    @classmethod
+    def tearDownClass(cls):
+        pass
+
+    def assertTupleAlmostEqual(self, first, second, places=None, msg=None, delta=None):
+        """
+        check two tuples are almost equal (value by value)
+        """
+        for f, s in zip(first, second):
+            self.assertAlmostEqual(f, s, places=places, msg=msg, delta=delta)
+
+    def test_read_write(self):
+        vals = [27000, 37000, 20000, 44000]
+        self.bc._write_registers(vals)
+        ret = self.bc._read_registers()
+        self.assertTupleAlmostEqual(vals, list(ret), places=1)
+
+    def test_set_shift(self):
+        startt = time.time()
+        shift = (5e-9, 8e-9)
+        self.bc.shift.value = shift
+        self.assertLess(time.time() - startt, 0.01, "Reading/writing took more than 10 ms.")
+        logging.debug("Shift value set to %s", self.bc.shift.value)
+        self.assertTupleAlmostEqual(self.bc.shift.value, shift, places=9)
+
```

Comment:
Please add a test to check nothing crazy happens when you put very large shift value. It should be clipped to the value corresponding to the device maximum.